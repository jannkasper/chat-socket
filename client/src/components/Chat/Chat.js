import React, { Component } from "react";

export class Chat extends Component {
    constructor(props) {
        super(props);
        this.state = {
            message: '',
            shiftKeyDown: false,
        };
    }

    componentDidMount() {
        // Disable for now due to vary issues:
        // Paste not working, shift+enter line breaks
        // autosize(this.textInput);
        this.textInput.addEventListener('autosize:resized', () => {
            this.props.scrollToBottom();
        });

    }

    handleFormSubmit(evt) {
        evt.preventDefault();
        this.sendMessage();
    }

    canSend() {
        return this.state.message.trim().length;
    }

    sendMessage() {
        if (!this.canSend()) {
            return;
        }

        const { message } = this.state;
        // const isCommand = this.parseCommand(message);

        // if (isCommand) {
        //     const res = this.executeCommand(isCommand);
        //     if (res === false) {
        //         return;
        //     }
        // } else {
        //     this.props.sendEncryptedMessage({
        //         type: 'TEXT_MESSAGE',
        //         payload: {
        //             text: message,
        //             timestamp: Date.now(),
        //         },
        //     });
        // }

        this.props.sendEncryptedMessage({
            type: 'TEXT_MESSAGE',
            payload: {
                text: message,
                timestamp: Date.now(),
            },
        });

        this.setState({
            message: '',
        });
    }

    handleInputChange(evt) {
        this.setState({
            message: evt.target.value,
        });
    }

    handleKeyUp(e) {
        if (e.key === 'Shift') {
            this.setState({
                shiftKeyDown: false,
            });
        }
    }

    handleKeyPress(e) {
        if (e.key === 'Shift') {
            this.setState({
                shiftKeyDown: true,
            });
        }
        // Fix when autosize is enabled - line breaks require shift+enter twice
        if (e.key === 'Enter' && !this.state.shiftKeyDown) {
            e.preventDefault();
            if (this.canSend()) {
                this.sendMessage();
            } else {
                this.setState({
                    message: '',
                });
            }
        }
    }

    render() {
        return (
            <form onSubmit={this.handleFormSubmit.bind(this)} className="chat-preflight-container">
                <textarea
                    rows="1"
                    onKeyUp={this.handleKeyUp.bind(this)}
                    onKeyDown={this.handleKeyPress.bind(this)}
                    ref={input => {
                        this.textInput = input;
                    }}
                    autoFocus
                    className="chat"
                    value={this.state.message}
                    placeholder={this.props.translations?.typePlaceholder}
                    onChange={this.handleInputChange.bind(this)}
                />
            </form>
            )
    }
}

export default Chat
