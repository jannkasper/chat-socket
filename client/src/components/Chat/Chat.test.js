import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import { Chat } from './Chat';

const translations = {
  typePlaceholder: 'inputplaceholder',
};

describe('Chat component', () => {

  beforeEach(() => {
    // Fake date
    jest.spyOn(global.Date, 'now').mockImplementation(() => new Date('2020-03-14T11:01:58.135Z').valueOf());
  })

  it('should display', () => {
    const {asFragment} = render(
        <Chat
            scrollToBottom={() => {}}
            // focusChat={false}
            // userId="foo"
            // username="user"
            // showNotice={() => {
            // }}
            // clearActivities={() => {
            // }}
            sendEncryptedMessage={() => {}}
            translations={{}}
        />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('can send message', () => {
    const sendEncryptedMessage = jest.fn();

    render(
        <Chat
            scrollToBottom={() => {
            }}
            // focusChat={false}
            // userId="foo"
            // username="user"
            // showNotice={() => {}}
            // clearActivities={() => {}}
            sendEncryptedMessage={sendEncryptedMessage}
            translations={translations}
        />,
    );

    const textarea = screen.getByPlaceholderText(translations.typePlaceholder);

    // Validate but without text
    fireEvent.keyDown(textarea, {key: 'Enter'});

    expect(sendEncryptedMessage).not.toHaveBeenCalled();

    // Type test
    fireEvent.change(textarea, {target: {value: 'test'}});
    // Validate
    fireEvent.keyDown(textarea, {key: 'Enter'});

    expect(sendEncryptedMessage).toHaveBeenLastCalledWith({
      payload: {text: 'test', timestamp: 1584183718135},
      type: 'TEXT_MESSAGE',
    });

    // Validate (textarea should be empty)
    fireEvent.keyDown(textarea, {key: 'Enter'});

    expect(sendEncryptedMessage).toHaveBeenCalledTimes(1);
  });

  it("shouldn't send message with Shift+enter", () => {
    const sendEncryptedMessage = jest.fn();

    render(
        <Chat
            scrollToBottom={() => {
            }}
            // focusChat={false}
            // userId="foo"
            // username="user"
            // showNotice={() => {}}
            // clearActivities={() => {}}
            sendEncryptedMessage={sendEncryptedMessage}
            translations={translations}
        />,
    );

    const textarea = screen.getByPlaceholderText(translations.typePlaceholder);

    // Test shift effect
    fireEvent.change(textarea, {target: {value: 'test2'}});
    fireEvent.keyDown(textarea, {key: 'Shift'});
    fireEvent.keyDown(textarea, {key: 'Enter'});
    fireEvent.keyUp(textarea, {key: 'Shift'});

    expect(sendEncryptedMessage).toHaveBeenCalledTimes(0);

    // Now we want to send the message
    fireEvent.keyDown(textarea, {key: 'Enter'});

    expect(sendEncryptedMessage).toHaveBeenCalledTimes(1);

    expect(sendEncryptedMessage).toHaveBeenLastCalledWith({
      payload: {text: 'test2', timestamp: 1584183718135},
      type: 'TEXT_MESSAGE',
    });
  });

});
