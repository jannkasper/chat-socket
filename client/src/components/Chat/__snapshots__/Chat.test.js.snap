// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat component should display 1`] = `
<DocumentFragment>
  <form
    class="chat-preflight-container"
  >
    <textarea
      class="chat"
      rows="1"
    />
  </form>
</DocumentFragment>
`;
